openpgp/packet/packet.go

```diff
@@ -356,10 +356,10 @@ func Read(r io.Reader) (p Packet, err error) {
 		if contents, version, err = peekVersion(contents); err != nil {
 			return
 		}
-		if version < 4 {
-			p = new(SignatureV3)
-		} else {
+		if version == 4 {
 			p = new(Signature)
+		} else {
+			err = errors.UnsupportedError("signature version")
```

Comment:
You actually need `contents` from `peekVersion` (which is the actual content of the packet, minus the first (version) byte. I had tried with
```
if contents, _, err = peekVersion(contents); err != nil {
	return
}
```
which works without signature validation (I didn't quite like it semantically). I can roll back to that.

Reply:
`peekVersion` is not ditching the first byte, it's only peeking at it, so the above is a no-op, and you can just use the `contents` as-is.

Reply:
I see. For some reason, not using `peekVersion` an using `contents` as is breaks some tests (multiple DSA sigs and MissingHashFunction which uses RIPEMD160).

Reply:
In other words, the instruction
```
contents = bufio.NewReader(contents)
```
is necessary here somehow (?) (passes all tests with this).

Reply:
I see, interesting. In that case, could you perhaps try to find out why/where it's necessary, and then place that line closer to where it's needed (or change the code there so that it's not needed)?

Reply:
After some effort I have more information about this. When parsing the signature
```
8a00000040040001030006050253863c24000a09103b4fe6acc0b21f32ffff0101010101010101010101010101010101010101010101010101010101010101010101010101
```
(used in `TestMissingHashFunction`, `TestMultipleSignaturePacketsDSA` and last modified [here](https://go-review.googlesource.com/c/crypto/+/13781/3/openpgp/read_test.go)) the trailing `0x01` bytes are interpreted as an RSA MPI. The MPI announces a bit length of
```
0x01 << 8 | 0x01 = 257 bits = (257 + 7)/8 bytes = 33 bytes.
```
In other words, there should be `2 + 33` bytes in the MPI. Now, there are 38 trailing `0x01` bytes, i.e., 3 extra bytes. On the other hand, the packet length is announced as 64 bytes, and indeed there are 64 bytes in the signature (the 69 bytes displayed above, minus the tag byte and the four bytes determining the length).

In a call to `Next` on a reader `r` of the above signature: At this point, a `spanReader` with 64 bytes count is initialized to read the first (and only) packet, and when parsing the MPI, it fails to consume the 3 extra `0x01` bytes because the MPI says that there are only 33. Because `contents` points to the span reader that points to the original packet reader `r`, the next call on `Next` considers these bytes as part of a new packet, and fails to read the header (`0x01` is invalid; MSB not set). In other words, `r` is at the wrong position (and rightly so, as per the MPI length).

Note that `peekVersion` (introduced [here](https://codereview.appspot.com/12746046/patch/51001/33002), finally, this PR is reverting that one) does not only peek the version byte, but also casts the input reader into a `bufio.Reader`, potentially changing the behavior (to be fair, this is documented above `peekVersion`). This also has the consequence that it reads all contents from the input reader and puts them in the new reader.

Now, the act of casting `contents` into something else means reading these 3 extra bytes from `r` and losing the reference (therefore the overlying packets reader `r` is at the right position dictated by the packet length and not the MPI length) and that's why the test passes if we cast `contents` into a `bufio.Reader` (i.e. calling `peekVersion`), `bytes.Reader`, or artificially reading those 3 bytes.

We now know that the above signature is not length consistent (assuming the MPI is not encrypted, why would it be), but there also is no check on the packet length after reading a signature packet.

`RSASignature.ReadFrom` should return a length error so that `consumeAll` could kick in to avoid these kind of problems (see also the TODO mark in packet/reader.go). Also, the error from `RSASignature.ReadFrom` is NOT checked.  This is easier said than done, since `ReadFrom` should accept malformed MPIs if they are encrypted (I dislike this).

On last commit:
- Remove `peekVersion`
- Check error from `RSASignature.ReadFrom`
- Test vector is corrected, and retain the old test vector as  `rsaSignatureBadMPILength`
- From now on, assume that MPIs are correctly formed in RSA signatures (`RSASignature.ReadFrom` returns a length error) and add a test for this behavior.

Reply:
I see. Thanks a lot for looking into this!

> This is easier said than done, since ReadFrom should accept malformed MPIs if they are encrypted (I dislike this).

Yes, I agree. MPIs in a signature should never be encrypted, and even if they were, we shouldn't be parsing them as MPIs in that case. For example this code (in private_key.go) also only parses MPIs when they are unencrypted:

https://github.com/ProtonMail/crypto/blob/00945d66b16dfa191e6d3c19b7f07e46b9c79181/openpgp/packet/private_key.go#L205-L212

> From now on, assume that MPIs are correctly formed in RSA signatures

:+1: